feat: add obsolete snapshot reporting

patrickhulce opened the pull request:
Discussion point for https://github.com/americanexpress/jest-image-snapshot/issues/59

This PR adds a reporter that automatically deletes obsolete snapshots.

~~Example output of this PR.~~ **This output was from an earlier version of this PR**

```
$ JEST_IMAGE_SNAPSHOT_TRACK_OBSOLETE=1 npm test -- --reporters='<rootDir>/src/obsolete-reporter.js'
{ touchedFiles:
  [ './__tests__/__image_snapshots__/integration-spec-js-to-match-image-snapshot-happy-path-matches-an-identical-snapshot-1-snap.png',
    './__tests__/__image_snapshots__/integration-6-snap.png',
    './__tests__/__image_snapshots__/integration-update-snap.png' ],
 allFiles:
  [ './__tests__/__image_snapshots__/integration-6-snap.png',
    './__tests__/__image_snapshots__/integration-obsolete-6-snap.png',
    './__tests__/__image_snapshots__/integration-spec-js-to-match-image-snapshot-happy-path-matches-an-identical-snapshot-1-snap.png',
    './__tests__/__image_snapshots__/integration-update-snap.png' ],
 imageSnapshotDirectories:
  [ './__tests__/__image_snapshots__' ],
 obsoleteFiles:
  [ './__tests__/__image_snapshots__/integration-obsolete-6-snap.png' ] }
```

Proposed end-user usage

```json
{
  "jest": {
    "reporters": ["default", "jest-image-snapshot/src/obsolete-reporter.js"]
  }
}
```

```bash
export JEST_IMAGE_SNAPSHOT_TRACK_OBSOLETE=1
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/americanexpress/jest-image-snapshot?pullRequest=222) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/americanexpress/jest-image-snapshot?pullRequest=222) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/americanexpress/jest-image-snapshot?pullRequest=222) it.</sub>

patrickhulce commented on src/obsolete-reporter.js:
```diff
@@ -0,0 +1,64 @@
+const fs = require("fs");
+const path = require("path");
+
+const TOUCHED_FILE_LIST_PATH = path.join(
```
this file needs to be something that can be determined within... 

- `toMatchImageSnapshot`
- `onRunStart`
- `onRunComplete`

Ideally I think this would live in the cache directory but `onRunStart` and `toMatchImageSnapshot` don't have obvious ways to access that. `process.cwd()` is the only thing I could think of that doesn't require config gymnastics.

patrickhulce commented on src/obsolete-reporter.js:
```diff
@@ -0,0 +1,64 @@
+  ".jest-image-snapshot-touched-files"
+);
+
+const IS_ENABLED = !!process.env.JEST_IMAGE_SNAPSHOT_TRACK_OBSOLETE;
```
gated behind a ENV flag

Kjens93 replied:
I’d like this flag to be opt-out, rather than opt-in. I think the majority of users will likely want to have it on by default. Just a thought.

patrickhulce replied:
While I agree the majority of users will want to have it on, it requires explicit usage of a custom reporter configured in jest that will be fundamentally optional. the reporter might be able to flip it on by default once it's included but I haven't looked into that yet (not sure if jest will respect whatever reporters do to `process.env` when spawning the workers)

Kjens93 replied:
@patrickhulce Totally understood. I'm not very familiar with how Jest's test reporter system works, but looking at this code, it appears that nothing in this file will ever even be executed unless the user adds `jest-image-snapshot/src/obsolete-reporter.js` to their Jest configuration, correct? In which case it feels like the environment variable is redundant.

patrickhulce replied:
> nothing in this file will ever even be executed unless the user adds jest-image-snapshot/src/obsolete-reporter.js to their Jest configuration, correct?

Correct.

> I probably wouldn't add the environment variable gate at all if it were me, but I'd settle for using it as an opt-out escape hatch.

AFAIK, it is intentionally impossible to tell within each jest worker at the level of `toMatchImageSnapshot()` what reporter is being used. Because of that, we need *some* way of signaling to the worker processes that it needs to write to this "touched files" list. `process.env` is a straightforward, safe option.

Alternatives I see:

- Keep this environmental variable internal only and enable it by default (already mentioned, not sure if jest will respect whatever reporters do to process.env when spawning the workers)
- Treat the very existence of the magic file name as a signal to write to it. Prone to cleanup issues, but I'd favor this approach if we decided to make this enabled by default.

anescobar1991 replied:
I think we can start with opt out and then if it works well and we can figure out how to have it cleanly enabled by default then we can make it a default. Easier to add than to remove features ;)

patrickhulce commented on src/obsolete-reporter.js:
```diff
@@ -0,0 +1,64 @@
+      fs.unlinkSync(TOUCHED_FILE_LIST_PATH);
+    }
+
+    console.log({
```
could automatically remove if set to update,

patrickhulce replied:
the original point was to just print the outdated files for the user to remove them manually, so yes a log would be necessary in that case.

for now, I've updated to automatically remove the files, but that feels slightly riskier IMO.

patrickhulce commented:
Anyone else with privileges here, is this something that is worth pursuing? If there is no intention of discussion or review here, it's not hurting any feelings, I'd just like clarity so as to not continue investing time and focus on a dead end.

cc @JamesSingleton I believe you asked to open a PR for a discussion (https://github.com/americanexpress/jest-image-snapshot/issues/59#issuecomment-652769176)

anescobar1991 commented:
Hey @patrickhulce sorry about that let me look at it on Monday.

anescobar1991 commented on src/index.js:
```diff
@@ -178,7 +179,8 @@ function configureToMatchImageSnapshot({
     const diffDir = customDiffDir || path.join(snapshotsDir, '__diff_output__');
     const baselineSnapshotPath = path.join(snapshotsDir, `${snapshotIdentifier}-snap.png`);
-
+    ObsoleteReporter.markTouchedFile(baselineSnapshotPath);
```
Can we add documentation about this in the readme?

Would be nice to have an integration test but can see why it would be challenging to do.

patrickhulce replied:
Sure 👍

anescobar1991 commented on src/obsolete-reporter.js:
```diff
@@ -0,0 +1,64 @@
+
+  static markTouchedFile(filePath) {
+    if (!IS_ENABLED) return;
+    const fd = fs.openSync(TOUCHED_FILE_LIST_PATH, "as");
```
`fd`? Can we use a more friendly name?

patrickhulce replied:
`fd` is fairly common for a file descriptor in my experience, but I can understand it's opaque for those that haven't seen the pattern before.

`touchedListFileDescriptor`?

patrickhulce commented on src/index.js:
```diff
@@ -18,6 +18,7 @@ const path = require('path');
 const Chalk = require('chalk').constructor;
 const { diffImageToSnapshot, runDiffImageToSnapshot } = require('./diff-snapshot');
 const fs = require('fs');
+const ObsoleteReporter = require('./obsolete-reporter');
```
😆 Sounds good, any suggestions? `SnapshotRemovalReporter`?

anescobar1991 replied:
outdated-snapshot-reporter?

patrickhulce commented:
Updated with feedback @anescobar1991 and added an end-to-end test for this but I'm not sure what to do about coverage since the statements being covered are in child processes that aren't instrumented. Any guidance here?

anescobar1991 commented on src/outdated-snapshot-reporter.js:
```diff
@@ -0,0 +1,89 @@
+
+class OutdatedSnapshotReporter {
+  /* istanbul ignore next - test coverage in child process */
+  static isEnabled() {
```
Is this used? I assume linting is running on this file can you make sure? I thought it would have caught this...

patrickhulce replied:
you're right this isn't being used anymore.

linting is running I had to ignore `class-methods-use-this` because to satisfy the jest reporter API

patrickhulce commented on src/outdated-snapshot-reporter.js:
```diff
@@ -0,0 +1,89 @@
+const IS_ENABLED = !!process.env.JEST_IMAGE_SNAPSHOT_TRACK_OBSOLETE;
+
+class OutdatedSnapshotReporter {
+  /* istanbul ignore next - test coverage in child process */
+  static isEnabled() {
+    return IS_ENABLED;
+  }
+
```
```suggestion
```
